### ISSUE TITLE
unable to drop read-only PG using `pg_drop_source` API

### ISSUE THREAD

**Issue by tirumaraiselvan** [hasura/graphql-engine#7092]
```
curl 'http://localhost:8080/v1/metadata' -d '{"type":"pg_drop_source","args":{"name":"read-only"}}'

{"internal":{"statement":"BEGIN ISOLATION LEVEL READ COMMITTED READ WRITE","prepared":true,"error":{"exec_status":"FatalError","hint":null,"message":"cannot set transaction read-write mode during recovery","status_code":"0A000","description":null},"arguments":[]},"path":"$.args","error":"postgres tx error","code":"postgres-error"}
```

@evertedsphere said this happens because we run [postDropSourceHook](https://github.com/hasura/graphql-engine-mono/blob/9a07fec31fcdbe6ef3b96bb61c95e0dd544e7f50/server/src-lib/Hasura/Backends/Postgres/DDL/Source.hs#L239) in read-write mode always.

**Workaround**: We can remove the read-only source by manually removing it from source list in metadata file and then re-applying it. This doesn't run the postDropSourceHook and succeeds. We should discuss these different behaviours when removing sources as well.

**Comment by tirumaraiselvan** [comment]
The fact that we don't cleanup (via `postDropSourceHook`) when a source is dropped via `replace_metadata`  is not correct. Consider, that we had event triggers on a source and then we removed it via metadata. Then, the source may still have pending events which could start getting delivered when the source is added again (all you need is atleast one ET configured) and this will have unpredictable behaviour.

This also seems against our diff-ing logic of building schema cache. The diff should show that we removed a source and we should take appropriate action on the deleted source (similar to how removing an event trigger from metadata will clear the PG triggers on the source). If an error occurs during cleanup, then we should log a warning.

**Comment by evertedsphere** [comment]
i've rethought this, and i agree now because it makes sense to think of `replace_metadata` not as a "manual update", but rather as a "bulk update" whose effects should be equivalent to updating each piece of the metadata one by one.

**Comment by evertedsphere** [comment]
i think this also points to the importance of making sure we clearly **define semantics or specs for api operations in our documentation** and describe what they _mean_ at a conceptual level, on top of our current practice of having docs that describe how the implementation works and what users can expect from the current version, so that when we later have to make deep or wide-ranging changes we have something to guide us. (cc @paf31)

additionally, we should also have similar specs for more abstract things like "metadata", "event trigger", and so on.

for instance, we could have a spec for metadata which mentions something like:

> graphql-engine stores some information known as _metadata_ in certain internal tables. graphql-engine considers the metadata its internal model of the world, storing information about connections, tables, stored procedures, sources, replicas, remote schemas, and so on. if the metadata is found to not match the environment in some way — for example, a tracked table is nonexistent, or a remote schema is down — graphql-engine throws a _metadata inconsistency error_.
>
>  whenever the metadata is changed in any way by a user, and if the change is of the following specified forms, graphql-engine will attempt to update the part of the environment it controls to match the new metadata in the specified way:
> 
> * creating the first event trigger on a read-write source (triggers creation of the `event_log` table)
> * dropping the only event trigger on a read-write source (triggers dropping of `event_log`)
>
> in case of a failure to make such a change, a metadata inconsistency error is thrown.

**Comment by ecthiender** [comment]
The conclusion here was that we run `postDropSourceHook` and log if there are any errors and continue, instead of error-ing out when there are errors from `postDropSourceHook`. That way, this will work on read-only databases.

The questions of if we should separate out the event catalog lifecycle management, and diff-ing during `replace_metadata` to run clean up actions can be solved later as a separate issue/task.

**Comment by ecthiender** [comment]
Fixed via hasura/graphql-engine-mono#1724
